Document let else statements (#1156)
rust-lang/reference: src/statements.md

> **<sup>Syntax</sup>**\
> _LetStatement_ :\
> &nbsp;&nbsp; [_OuterAttribute_]<sup>\*</sup> `let` [_PatternNoTopAlt_]
>     ( `:` [_Type_] )<sup>?</sup> ( | `=` [_Expression_] | `=` [_Expression_]

> >     `else` [_BlockExpression_]) `;`

Author: Do you know how to do the note in source code?

> >     `else` [_BlockExpression_]) `;`

Author: (also yes the leading | was to indicate the epsilon case, but I have changed it to your suggestion)

> >     `else` [_BlockExpression_]) `;`

Reviewer: Oh, sorry, I should have included the source from the comment above:

```
> **<sup>Syntax</sup>**\
> … [_Expression_] [†](#let-else-restriction) …
> 
> <span id="let-else-restriction">† When an `else` block is specified, the _Expression_ must not be a [_LazyBooleanExpression_](expressions/operator-expression.md#lazy-boolean-operators) or end with a `}`.</span>
```

Just make it a link, and use a `<span>` to set the target.

> >     `else` [_BlockExpression_]) `;`

Author: Thanks! I've tried out your code, it causes an error in CI:

> error in ../src/statements.md (line 57): broken Reference link (reference `†`)
some style checks failed

Apparently pulldown-cmark does not parse the inline html.

I have tried using a footnote, but that shows a <sup>1</sup>, which in this context can be very easily misunderstood as a repetition:
![Screenshot_20220924_031548](https://user-images.githubusercontent.com/8872119/192074031-572317e2-d44f-4681-9f42-4ceeb6db9804.png)

However, it is possible to hide the number and put a † on its stead. I've pushed a commit to do that.

> >     `else` [_BlockExpression_]) `;`

Reviewer: I would prefer to not change the behavior of footnotes. Those are used in other places of the reference.

I would also prefer to avoid using them in this situation. This isn't really a footnote, but more of an attempt to stuff a bunch of information in a small amount of space.

I tried the suggestion I posted above, and the `style-check` and the linkchecker both passed.  It should look exactly like:

```md
> **<sup>Syntax</sup>**\
> _LetStatement_ :\
> &nbsp;&nbsp; [_OuterAttribute_]<sup>\*</sup> `let` [_PatternNoTopAlt_]
>     ( `:` [_Type_] )<sup>?</sup> (`=` [_Expression_] [†](#let-else-restriction)
>     ( `else` [_BlockExpression_]) <sup>?</sup> ) <sup>?</sup> `;`
>
> <span id="let-else-restriction">† When an `else` block is specified, the _Expression_ must not be a [_LazyBooleanExpression_](expressions/operator-expr.md#lazy-boolean-operators) or end with a `}`.</span>
```

> >     `else` [_BlockExpression_]) `;`

Author: @ehuss regarding the CI failure, please see [this](https://github.com/rust-lang/reference/actions/runs/3116465400/jobs/5054319422) CI run, ran from [this](https://github.com/rust-lang/reference/pull/1156/commits/3eacf0029c5298c481fca9db1e1adf5d313ed293) commit. If you click on the prev button you will get to the commit that has precisely the state that you suggest.

```
    Compiling style-check v0.1.0 (/home/runner/work/reference/reference/style-check)
    Finished dev [unoptimized + debuginfo] target(s) in 37.63s
     Running `target/debug/style-check ../src`
error in ../src/statements.md (line 57): broken Reference link (reference `†`)
some style checks failed
```

```markdown
> &nbsp;&nbsp; [_OuterAttribute_]<sup>\*</sup> `let` [_PatternNoTopAlt_]
>     ( `:` [_Type_] )<sup>?</sup> (`=` [_Expression_] )<sup>?</sup> `;`
>     ( `:` [_Type_] )<sup>?</sup> (`=` [_Expression_][†](#let-else-restriction)
>     ( `else` [_BlockExpression_]) <sup>?</sup> ) <sup>?</sup> `;`
A *`let` statement* introduces a new set of [variables], given by an
<span id="let-else-restriction">† When an `else` block is specified, the
irrefutable [pattern]. The pattern is followed optionally by a type
_Expression_ must not be a [_LazyBooleanExpression_], or end with a `}`.</span>
```

I can try it again, but I doubt it will be any different.

> >     `else` [_BlockExpression_]) `;`

Reviewer: That's because in that commit you have:

```
[_Expression_][†](#let-else-restriction)
```

The text I have above has a space:

```
[_Expression_] [†](#let-else-restriction)
```

Without the space, markdown thinks the `[†]` is a reference link for the preceding `[_Expression_]`.

> >     `else` [_BlockExpression_]) `;`

Author: Ohhhh... I see... Will adjust it right away.

A *`let` statement* introduces a new set of [variables], given by a [pattern].
The pattern is followed optionally by a type annotation and then either ends,
or is followed by an initializer expression plus an optional `else` block.
When no type annotation is given, the compiler will infer the type, or signal
an error if insufficient type information is available for definite
inference. Any variables introduced by a variable declaration are visible
from the point of declaration until the end of the enclosing block scope,
except when they are shadowed by another variable declaration.

If an `else` block is not present, the pattern has to be irrefutable.

> If an `else` block is present, the pattern may be refutable, and if it is, it
> will be checked on run time. On a mismatch, the `else` block is executed.

Reviewer: I feel like specifying "run time" here may be over-specifying the behavior. For example, a `let` in a `const` will be evaluated at compile-time. It also isn't specifically stated what it means "to check", which isn't a verb typically used in the reference with pattern matching. These seem like two clauses to me, so I think it would be fine to state them separately, such as:

```suggestion
If an `else` block is present, the pattern may be refutable.
If the pattern does not match the initializer expression, the `else` block is executed.
```

Author: const-evaluation still has a "run-time" and compile time separation. But shrug.

The `else` block has to always diverge (evaluate to type `!`).

> If an `else` block is present, restrictions apply on the expression:
> It might not be a [_LazyBooleanExpression_], or end in a `}` token.

Reviewer: I think this should be up in the grammar section since it is part of the grammar.

Also, similar to the must/should/may terminology:

```suggestion
If an `else` block is present, restrictions apply on the expression:
It must not be a [_LazyBooleanExpression_], or end in a `}` token.
```

Although I do lean towards the earlier wording I suggested. I think it can sometimes be a little difficult to follow when using "it" when it is not very clear what "it" is. Here it is referring to an "expression", but there are different expressions at play (such as the block expression, and all the expressions within the block). Up in the grammar, the word _Expression_ refers to a specific expression (there is only one of that form; that is a rule with a capital E and italicized).

Author: I think it's clear because there is only one expression in the grammar. It is also mentioned in the first paragraph.

> **<sup>Syntax</sup>**\
> _LetStatement_ :\
> &nbsp;&nbsp; [_OuterAttribute_]<sup>\*</sup> `let` [_PatternNoTopAlt_]
>     ( `:` [_Type_] )<sup>?</sup> (`=` [_Expression_] [†](#let-else-restriction)
>     ( `else` [_BlockExpression_]) <sup>?</sup> ) <sup>?</sup> `;`

> <span id="let-else-restriction">† When an `else` block is specified, the

Author: Done

from the point of declaration until the end of the enclosing block scope,
except when they are shadowed by another variable declaration.

### `let else` statements

> **<sup>Syntax</sup>**\
> _LetElseStatement_ :\
> &nbsp;&nbsp; [_OuterAttribute_]<sup>\*</sup> `let` [_PatternNoTopAlt_]
>     ( `:` [_Type_] )<sup>?</sup> `=` [_Expression_] `else` [_BlockExpression_] `;`

A *`let else` statement* introduces a new set of [variables], given by a

> refutable [pattern]. The pattern is followed optionally by a type

Reviewer 2: Is this true for nested type values? I am a little unsure how the type annotation feature works

```rust
// If it were this, this documentation statement would be true:
let Ok(v): Result<u32, _> = expr else { };

// _I thought_ it as supposed to be able to be like this though, but maybe I am wrong:
let Ok(v: u32) = expr else { };
```

Reviewer 3: No, let-else has the same type annotation syntax as let, not sure where you got that idea from.

Reviewer 2: Oh sorry I thought that was part of the upcoming "type ascription" for some reason.

from the point of declaration until the end of the enclosing block scope,
except when they are shadowed by another variable declaration.

### `let else` statements

> **<sup>Syntax</sup>**\
> _LetElseStatement_ :\
> &nbsp;&nbsp; [_OuterAttribute_]<sup>\*</sup> `let` [_PatternNoTopAlt_]
>     ( `:` [_Type_] )<sup>?</sup> `=` [_Expression_] `else` [_BlockExpression_] `;`

A *`let else` statement* introduces a new set of [variables], given by a
[pattern]. The pattern is followed optionally by a type annotation. When no type
annotation is given, the compiler will infer the type, or signal an error if
insufficient type information is available for definite inference. It is is
then followed by an initializer expression as well as a diverging `else` block.
Compared to `let` statement patterns which have to be irrefutable, `let else`

> patterns can be both irrefutable and refutable. In the case of irrefutable

Reviewer 4: ```suggestion
patterns can be either irrefutable or refutable. In the case of irrefutable
```

>     ( `:` [_Type_] )<sup>?</sup> `=` [_Expression_] `else` [_BlockExpression_] `;`

A *`let else` statement* introduces a new set of [variables], given by a
[pattern]. The pattern is followed optionally by a type annotation. When no type
annotation is given, the compiler will infer the type, or signal an error if
insufficient type information is available for definite inference. It is is
then followed by an initializer expression as well as a diverging `else` block.
Compared to `let` statement patterns which have to be irrefutable, `let else`
patterns can be both irrefutable and refutable. However, as it is besides the

> point of, irrefutable patterns in `let else` are linted against.

Reviewer 2: `as it is besides the point of` reads really strange...

Something like this would be better:
```
However, irrefutable patterns in `let else` are linted against due to their lack of usefulness (the `else` would be dead code).
```

Author: What about "Irrefutable patterns are not forbidden, however there is a lint for them as the `else` clause is redundant"?

Reviewer 2: That sounds better.

Author: I've applied a slightly different version, but one that avoids the "besides the point" formulation.

Reviewer 2: Looks good to me
